File: 85-maximal-rectangle/maximal-rectangle.cpp
class Solution {
public:
    int mah(int n,vector<int> &a)
    {
        int ans=0;
        stack<int> s;
        vector<int> v;
         for(int i=n-1;i>=0;i--)
        {
            while(!s.empty()&&a[s.top()]>=a[i])
                s.pop();
            if(s.empty())
                v.push_back(n);
            else
                v.push_back(s.top());
        s.push(i);
         }
        reverse(v.begin(),v.end());
  
        while(!s.empty())
            s.pop();
        
         for(int i=0;i<n;i++)
         {
              while(!s.empty()&&a[s.top()]>=a[i])
                s.pop();
             
                if(s.empty())
                v[i]=v[i]*a[i];
            else
                v[i]=(v[i]-s.top()-1)*a[i];
             
             s.push(i);
             
             ans=max(ans,v[i]);
        }
        return ans;
    }
    int maximalRectangle(vector<vector<char>>& b) {
        int ans=0,n=b.size(),m=b[0].size(),i=0;
        vector<int> a(m);
        while(i<n)
        {
            for(int j=0;j<m;j++)
            {
                if(b[i][j]=='0')
                    a[j]=0;
                else
                    a[j]+=1;
            }

            ans=max(ans,mah(m,a));
            i++;
        }
        return ans;
    }
};

File: 85-maximal-rectangle/maximal-rectangle.java
class Solution {
    int cal(int h[], int n)
    {
        Stack<Integer> s= new Stack<>();
        int ans=0;
        for(int i=0;i<n;i++)
        {
            while(!s.isEmpty() && h[s.peek()]>h[i])
            {
                int k=s.pop();
                int len = s.isEmpty() ? i : i-s.peek()-1;
                ans=Math.max(ans,len*h[k]);
            }
            s.push(i);
        }

            while(!s.isEmpty())
            {
                int k=s.pop();
                int len = s.isEmpty() ? n : n-s.peek()-1;
                ans=Math.max(ans,len*h[k]);
            }
        return ans;
    }
    public int maximalRectangle(char[][] matrix) {
        int m=matrix.length,n=matrix[0].length;
        int t[][]= new int[m][n];

        for(int j=0;j<n;j++)
            t[0][j]=matrix[0][j]-'0';
        
        for(int i=1;i<m;i++)
        for(int j=0;j<n;j++)
            if(matrix[i][j]=='1')
                t[i][j]=1+t[i-1][j];
        int ans=0;

        for(int i=0;i<m;i++)
            ans=Math.max(ans,cal(t[i],n));
        return ans;
    }
}